### DISCUSSION TITLE
TabView and long-running effects

### DISCUSSION THREAD

**Discussion by vykut** [pointfreeco/swift-composable-architecture#2655] [Q&A] [answered]
Hi,

In my application, I have a tab view with 3 screens, each one of them and the TabView having their own `Reducer`. 
In 2 of those reducers, I'm starting long-running effects via the `View.onAppear`, one which is an observation of database changes via `AsyncStream` (i.e `for recipe in database.recipes { await send(.updateRecipes(recipes) }`), while the other is a network request for a random recipe, at an interval of 30 seconds.

I'm noticing that every time I'm changing between tabs, say going from tab 1 to tab 3 and back to tab 1, these "onAppear" effects are canceled every time I leave their screen, i.e tab 1 -> tab 3 => reducer 1 cancels its currently running "onAppear" effect.

Is there a way to start an effect in a reducer, for its entire lifetime? i.e start an effect in the reducer's `init`
If there is not, how else should I handle these long-running effects?

I was thinking I could spawn a long-running effect for each tab, in the Tab View reducer (which doesn't trigger `onAppear` more than once) and `send` actions to each child reducer.

At the same time, I think the current behavior of reducers may actually be the correct one and in fact, these observations should be canceled once the user moves to a different tab and restarted when the user comes back. In that case, I'm not sure how I should handle a long-running effect that makes a network request based on a timer (i.e fetch a random recipe every 30 seconds and display it in the UI). In this case, every time the user switches back and forth between tabs, a random recipe is shown less than 30 seconds at a time.

**Comment by mbrandonw** [comment]
> I'm noticing that every time I'm changing between tabs, say going from tab 1 to tab 3 and back to tab 1, these "onAppear" effects are canceled every time I leave their screen, i.e tab 1 -> tab 3 => reducer 1 cancels its currently running "onAppear" effect.

This should not be the case if you are using `onAppear`. Any effect created from sending an action in `onAppear` is untethered from the presentation of the tab, and should continue executing regardless if it's visible or not.

Could you share a minimal demo project that shows the problem you are seeing?

**Reply by vykut** [reply]
Hi @mbrandonw,

You are right, in my app I was using the `.task` modifier, instead of the `.onAppear`, which gets canceled once the view "disappears" (i.e user switches to another tab).

However, this doesn't address my root issue, which is how can I make only *one* long-running effect, regardless of the presentation state of the view, or would this be an anti-pattern?
I could store a boolean in the `Reducer.State` i.e. `isRunningOnAppearEffect` and base the logic on it, combined with the "non-cancellable" `.onAppear` modifier. Are there better alternatives?

**Reply by mbrandonw** [reply, marked as answer]
Yes, you probably do not want to use `.task`. 

I personally would recommend creating your own kind of `onFirstAppear` view modifier that that acts like `onAppear`, but fires its closure only once. You would use some local `@State` in the modifier to determine that condition. You can also find versions of it online if you do not want to make your own.

I would not recommend storing that state in your feature. It just isn't that important for your feature's logic, and so is fine to let the view take care of it.
